runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/BeamFnMapTaskExecutorFactory.java

```diff
@@ -293,12 +333,45 @@ public Node typedApply(RemoteGrpcPortNode input) {
                   registerFnOperation::getProcessBundleInstructionId,
                   (Coder) coder,
                   context);
+          LOG.info("[BOYUANZ LOG]: grpc write op coder {}", coder);
         }
         return OperationNode.create(operation);
       }
     };
   }
 
+  private Function<Node, Node> createOperationTransformForExecutableStageNode(
+      final Network<Node, Edge> network,
+      final String stageName,
+      final DataflowExecutionContext<?> executionContext,
+      final JobBundleFactory jobBundleFactory) {
+    return new TypeSafeNodeFunction<ExecutableStageNode>(ExecutableStageNode.class) {
+      @Override
+      public Node typedApply(ExecutableStageNode input) {
+        StageBundleFactory stageBundleFactory =
+            jobBundleFactory.forStage(input.getExecutableStage());
+        Iterable<OutputReceiverNode> outputReceiverNodes =
+            Iterables.filter(network.successors(input), OutputReceiverNode.class);
+
+        OutputReceiver[] outputReceivers = new OutputReceiver[Iterables.size(outputReceiverNodes)];
+        int index = 0;
+        Iterator<OutputReceiverNode> iterator = outputReceiverNodes.iterator();
+        while (iterator.hasNext()) {
```

Comment:
I think there must be a method that turns iterator into an array.

Reply:
I checked the doc talking about how to convert an iterator into a list. I found the under layer of implementation is copy the whole content into a immutable list. Then using this converting method will have the same time complexity and extra memory required, vs iterating on the iterator directly. Is there any concern about iterating it directly?